server/coordinator/procedure/operation/transferleader/transfer_leader.go

```diff
@@ -106,22 +109,31 @@ func NewProcedure(params ProcedureParams) (procedure.Procedure, error) {
 	}, nil
 }
 
-func buildRelatedVersionInfo(params ProcedureParams) procedure.RelatedVersionInfo {
+func buildRelatedVersionInfo(params ProcedureParams) (procedure.RelatedVersionInfo, error) {
 	shardViewWithVersion := make(map[storage.ShardID]uint64, 0)
-	for _, shardView := range params.ClusterSnapShot.Topology.ShardViews {
-		if shardView.ShardID == params.ShardID {
-			shardViewWithVersion[params.ShardID] = shardView.Version
-		}
+	shardView, exists := params.ClusterSnapshot.Topology.ShardViewsMapping[params.ShardID]
+	if !exists {
+		return procedure.RelatedVersionInfo{}, errors.WithMessagef(metadata.ErrShardNotFound, "shard not found in topology, shardID:%d", params.ShardID)
 	}
+	shardViewWithVersion[params.ShardID] = shardView.Version
+
 	relatedVersionInfo := procedure.RelatedVersionInfo{
-		ClusterID:        params.ClusterSnapShot.Topology.ClusterView.ClusterID,
+		ClusterID:        params.ClusterSnapshot.Topology.ClusterView.ClusterID,
 		ShardWithVersion: shardViewWithVersion,
-		ClusterVersion:   params.ClusterSnapShot.Topology.ClusterView.Version,
+		ClusterVersion:   params.ClusterSnapshot.Topology.ClusterView.Version,
 	}
-	return relatedVersionInfo
+	return relatedVersionInfo, nil
 }
 
 func validateClusterTopology(topology metadata.Topology, shardID storage.ShardID, oldLeaderNodeName string) error {
+	_, found := topology.ShardViewsMapping[shardID]
+	if !found {
+		log.Error("shard not found", zap.Uint64("shardID", uint64(shardID)))
+		return metadata.ErrShardNotFound
+	}
+	if len(oldLeaderNodeName) == 0 {
+		return nil
```

Comment:
Isn't this an error for transfer_leader procedure?

Reply:
TransferLeader procedure ensure the OldLeaderNodeName is empty, which means that the shard has not been assigned to any node, which will appear when the cluster is initialized.